odl/tomo/operators/ray_trafo.py

```diff
@@ -273,164 +161,119 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
         else:
             # proj_space was given, checking some stuff
             if not isinstance(proj_space, DiscreteLp):
-                raise TypeError('`{}` must be a `DiscreteLp` instance, '
-                                'got {!r}'.format(proj_name, proj_space))
+                raise TypeError(
+                    '`proj_space` must be a `DiscreteLp` instance, '
+                    'got {!r}'.format(proj_space))
             if proj_space.shape != geometry.partition.shape:
-                raise ValueError('`{}.shape` not equal to `geometry.shape`: '
-                                 '{} != {}'.format(proj_name, proj_space.shape,
-                                                   geometry.partition.shape))
+                raise ValueError(
+                    '`proj_space.shape` not equal to `geometry.shape`: '
+                    '{} != {}'.format(proj_space.shape,
+                                      geometry.partition.shape))
             if proj_space.dtype != reco_space.dtype:
-                raise ValueError('`{}.dtype` not equal to `{}.dtype`: '
-                                 '{} != {}'.format(proj_name, reco_name,
-                                                   proj_space.dtype,
-                                                   reco_space.dtype))
+                raise ValueError('`proj_space.dtype` not equal to '
+                                 '`reco_space.dtype`: {} != {}'
+                                 .format(proj_space.dtype, reco_space.dtype))
+
+        if reco_space.ndim != geometry.ndim:
+            raise ValueError('`reco_space.ndim` not equal to `geometry.ndim`: '
+                             '{} != {}'.format(reco_space.ndim,
+                                               geometry.ndim))
+
+        # Cache for input/output arrays of transforms
+        self.use_cache = kwargs.pop('use_cache', True)
+
+        # Check `impl`
+        impl = kwargs.pop('impl', None)
+        impl_type, self.__cached_impl = self._check_impl(impl)
+        self._impl_type = impl_type
+        self.__impl = impl.lower() if isinstance(impl, str) else impl_type
+
+        self.geometry = geometry
```

Comment:
This should ultimately be a `@property`, too.

Reply:
I did this, but why is this exactly? For autocompletion?

Reply:
The main reason is to make our objects "publicly immutable". You can't assign to properties unless a setter is defined, whereas the plain attributes can be changed arbitrarily. It's not perfect but at least conveys an intention.

Reply:
That makes a lot of sense, thanks!